lib/ecto/query/planner.ex

```diff
@@ -236,13 +236,29 @@ defmodule Ecto.Query.Planner do
     end
   end
 
+  defp prepare_source(query, %FromExpr{source: source} = from, adapter),
+    do: %{from | source: prepare_source(query, source, adapter),
+                 prefix: prepare_prefix(query, from)}
   defp prepare_source(_query, {nil, schema}, _adapter) when is_atom(schema) and schema != nil,
     do: {schema.__schema__(:source), schema}
   defp prepare_source(_query, {source, schema}, _adapter) when is_binary(source) and is_atom(schema),
     do: {source, schema}
   defp prepare_source(_query, {:fragment, _, _} = source, _adapter),
     do: source
 
+  defp prepare_prefix(query, {_source, nil}),
+    do: query.prefix
+  defp prepare_prefix(query, {_source, schema}) when is_atom(schema) and schema != nil,
+    do: schema.__schema__(:prefix) || query.prefix
+  defp prepare_prefix(query, {:fragment, _, _}),
+    do: query.prefix
+  defp prepare_prefix(query, %FromExpr{} = from),
+    do: from.prefix || query.prefix
+  defp prepare_prefix(query, %JoinExpr{} = join),
+    do: join.prefix || query.prefix
+  defp prepare_prefix(query, %Ecto.SubQuery{} = subquery),
+    do: subquery.query.prefix || query.prefix
```

Comment:
I don't like that we are calling `prepare_prefix` recursively. This means if that if we accidentally put a `FromExpr` inside another FromExpr or if we forget to wrap a `from` inside a `FromExpr` at some point, it will go unnoticed.

Also, we don't need to fallback to `|| query.prefix` here as we ask the adapters to do that later on. Unless we want to consistently do so here... in this case, we should remove the `|| query.prefix` from the adapters. I assume normalizing it here is better as it requires less logic in the adapters?

Reply:
Same for `prepare_source` above.

Reply:
`prepare_prefix` isn't being called recursively.

Reply:
You are correct, only prepare_source above.